i18n replacement and pluralization

ron666 opened the pull request:
This PR creates a new class to translate and also integrates pluralization for strings following a new format for them. Solves #1792 and #1187

ron666 commented:
@OlivierJaquemet Please check this PR and let me know your thoughts. For the translations you created I'll need them to follow the format specified in English and Spanish NOT translating the word plural, given that is the name of the filter I'm using to pluralize. Refer to the rules inside the i18n file but you'll need to provide up to 3 different strings concatenated with `:`. Let me know if you have any questions/concerns

OlivierJaquemet commented on src/js/me-i18n.js:
```diff
@@ -1,169 +1,226 @@
-    }
-
-}(mejs.i18n.locale.strings));
+// ;(function (mejs, undefined) {
```
Maybe remove this whole section instead of commenting, then informat the Wordpress team of the modification ?

ron666 commented:
Thanks for reviewing this. About the first item, I'll see how can I integrate the rules inside the t method, since your arguments are valid. The Wordpress part I'll remove it. I could integrate a fallback but I think it will be better to integrate this right at once; so if you can assist with this I'd appreciate it. Let me know your thoughts

OlivierJaquemet commented:
Regarding the update of existing translation files to prevent use of a fallback : we do not have our own translators at Jalios, we rely on an external provider for this job. Because this service is quite expensive, but mainly due to the release cycle of our software, we only send the translation every 6/9 months or so. Unfortunately I already sent the previous strings to translation at the begining of the week... :( I'll see if I can request a modification with them and I'll keep you posted. 
But you should really consider providing this fallback in the meantime...

ron666 commented:
@OlivierJaquemet I'll add the fallback for now and let me know when you have the new translation available so we conclude this. I'll keep you posted. Thanks

ron666 commented:
@OlivierJaquemet Changes in place; please review and let me know. @ocean90 can you review this as well, please?

ron666 commented:
@OlivierJaquement I noticed an issue with re Brazilian Portuguese translation; shouldn't it be `exports[pt-BR]` instead of `exports.pt`?

ron666 commented:
And also as a reminder you need to check which languages would need 3 strings once you send them for an update; all the ideographic languages like Chinese or Japanese won't need it. Russian, Polish, etc. will need 3 and the romantic languages 2. Any questions let me know

OlivierJaquemet commented:
Brazilian portuguese should indeed follow all other translation format. see #1872
@odnamrataizem : you initially commited this pt-BR translation which is no longer up to date given the new strings that were introduced since your transalation. Could update your work starting from the english version please ?

@ron666 reminder regarding 3strings vs 2 duly noted.

OlivierJaquemet commented on src/js/me-i18n.js:
```diff
@@ -112,34 +112,101 @@
+		 * Translate a string to a specified language, including optionally a number to pluralize translation
 		 *
 		 * @param {String} input
+		 * @param {Number} pluralReplacement
```
Usually, arguments provided in a translation methods are understood by developpers as values being used in the translation message (exatly the simple use %1 was providing), indicating in the variable name that it is a number and it may also provides additionnal formating such as management of plural is a little bit confusing I think.
Maybe use a simpler and generic name like `msgParams`, `params`, `values`, `args` ?

Another possibility would be to omit this explicit variable declaration and use the `arguments` array (shifted by one to remove the input), like the original implementation of format method. It would allow any number of arguments and would not imply their type and feature (which are provided by filters). Of course this is more complex and not needed given the current use cases in mejs. So supporting only one message parameter is fine for me.

ron666 replied:
Thanks for pointing this. I'll change the argument names to `message` and `pluralParam`. Originally I had it with arguments, but like you stated it's not really needed for our purposes

ocean90 commented on src/js/me-i18n.js:
```diff
@@ -149,78 +216,8 @@
-	// i18n fixes for compatibility with WordPress
-	// if (typeof mejsL10n !== 'undefined') {
-	// 	i18n.locale.language = mejsL10n.language;
-	// }
```
What would be the new way to register translated strings?

ron666 replied:
I need to bring this back since this seems to be still needed

ron666 replied:
Wordpress support back in file; my apologies for removing that

ocean90 commented on src/js/me-i18n.js:
```diff
@@ -1,169 +1,223 @@
+		 * @see https://developer.mozilla.org/en-US/docs/Mozilla/Localization/Localization_and_Plurals#List_of_Plural_Rules
+		 * @type {Object[]}
+		 */
+		rules: [
```
This seems to be a bad idea because it hardcodes the languages and doesn't even include all languages.
I suggest that this data is stored in the translation file itself. The format which Gettext PO uses is quite popular. You can find a list of rules here: http://localization-guide.readthedocs.io/en/latest/l10n/pluralforms.html

The only downside: it seems to [require a lot of code](https://github.com/SlexAxton/Jed/blob/master/jed.js#L509-L1014).

Another idea is to include **all** plural forms and in the translation file you can define the # of a plural form which should be used.

ron666 replied:
For the purposes of the player it's overkill to try to use Jed or more complex libraries. Although I agree that we need to bring the rest of the languages so I'll add them later. The translation files must remain simple in my opinion; how would you include the plural number in the current format for a translation file?

ocean90 replied:
This could be a simple string like
```js
exports.en = {
    'mejs.plural-form': 'nplurals=2; plural=(n > 1);',
   // ...
}
```

or 

```js
exports.en = {
    'mejs.plural-form': '0', // value can be 0-16 if you use the Mozilla version or 1-6 (0-5) if you use http://unicode.org/cldr/trac/browser/trunk/common/supplemental/plurals.xml
   // ...
}
```

ron666 commented on src/js/me-i18n-locale-en.js:
```diff
@@ -33,7 +33,7 @@
 
             // mep-feature-jumpforward
-            'mejs.time-jump-forward': 'Jump forward %1 seconds',
+            'mejs.time-jump-forward': 'Jump forward {0} {0|plural:second:seconds}',
```
It will look `{0} {0|plural:X:Y:Z}` for Russian or Polish for example.

ron666 replied:
In your experience, what could be a more general format? The only thing I want to avoid is to overcomplicate this given that the translations we use with plurals are really minimal, but I also want to follow standards as much as possible

ron666 replied:
I changed this to `Jump forward 1 second;Jump forward %1 seconds`; that way we'll concatenate all the possible translations in the order dictated on http://localization-guide.readthedocs.io/en/latest/l10n/pluralforms.html

ocean90 replied:
What happens if a string or a translations has a semicolon? Shouldn't this be an array instead?

ron666 replied:
Yeah it's something I discussed and I think you are right. I'll use an array to account for this and adjust code; other than that are you Ok with this approach?

ron666 commented:
@OlivierJaquemet  and @ocean90 I refactored the method following the conventions of PluginForm.jsm using semi-colon concatenated strings to determine proper plural, and added all the plural forms to have a complete set of translations. Please review and let me know your thoughts

ron666 commented:
As a last pass I'll add documentation in the event someone wants to create a translation file. @OlivierJaquemet any comments on this that could be useful for other contributors? @ocean90 I hope with this new way it's more conventional and doesn't disturb the current support to WordPress, any more comments about this PR?

OlivierJaquemet commented:
No comment I can think of for the moment. Thanks again for your work.

ron666 commented:
@OlivierJaquemet Any comments on the approach of semi-colon strings, rather than using brackets? Do you think it is more standard this way?

OlivierJaquemet commented:
@ron666  I had not seen this modification (my apologies ... too much work here, not much time to properly review :/) 

If I understand correctly, you expect the whole message/sentence (and not just each parameter) to be proposed in several variations ?
- this is definitely not a standard I had seen (but again, I'm not an expert in I18N and properly management plurals for ALL languages is still a work to do for our translations at Jalios...) 
- it completely prevents the use of several parameters should you want to display a message like `Playlist contains {0,no videos|one video|{0} videos} and {1,no sounds|one sound|{1} sounds}.`
- for very long message, this adds a lot of duplicated text in the JS file

ron666 commented:
@OlivierJaquemet correct. The problem with that is that I saw on some languages is almost a completely new string and that in the long term would complicate things. Now, for the purposes of the player, and to maintain compatibility with WordPress, I think it's more likely to have something like the new approach and it won't be as confusing as the first approach I took. The highest number we can get is 6 translation (Arabic)

OlivierJaquemet commented:
@ron666 I'm fine with it. And it will definitely be simpler for translators (that would otherwise have to learn and understand complex syntax). How to do you insert a semicolons if you need one to be displayed to the end user ? (The answer can be : you don't :p)

ron666 commented:
@OlivierJaquemet yeah you don't but if it's needed I can change it by using something more unusual (like ~ or #) or simply doubling the semi-colons (;;)

ron666 commented:
@OlivierJaquemet I will change the semi-colon separated string to an array so we can account for semi-colons as @ocean90 pointed out

ron666 commented:
@OlivierJaquemet and @ocean90 I changed this to arrays and it will be even more clear and we won't have to worry about punctuation signs. Let me know your thoughts

OlivierJaquemet commented:
Arrays are good, thanks. 

One missing bit : when messages are inserted as HTML or HTML attribute, some characters must be encoded (as HTML entities) to output valid HTML. 
Two possibilities to address this : 
- encoding is integrated in the `t` function, eg `t("R&D") --> R&amp;D`
- another function wrapper is used each time t return value must be inserted in HTML `encodeFor...(t("R&D"))`

Obsviously, the first solution has the advantage of being applied globally to ensure proper encoding is never forgotten. the second solution allows for more use case (eg if I18N message but be printed in console, or any other plain text output), but they are probably not in use for now.

ron666 commented:
Thanks for pointing this. I'll add this later to complete this PR. Any other concerns? @ocean90 any thoughts or this is good to go once the HTML tags are escaped?

ron666 commented:
@OlivierJaquemet HTML escaping added. I'll wait for your final feedback as well as @ocean90 to merge this next week. Thanks.

ron666 commented:
Any objections if I merge this by tomorrow? I'm speeding this since this must be part of 3.0 and we are going to release it soon

OlivierJaquemet commented:
It looks good to me.

ron666 commented:
@OlivierJaquemet  and @ocean90 thanks for your help
